File: leetcode-python/weeklycontest/174/problem4.py
from collections import defaultdict, deque, Counter
from typing import List
import math
import copy
import random
import numpy as np
import inspect

# 题目没读懂
# bad code base 影响太大了

# TLE 实现不优雅


class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:
        dd = {}
        n = len(arr)
        for i in range(n-1):
            for j in range(i+1, n):
                if i + 1 == j:
                    dd[(i, j)] = min(arr[i], arr[j])
                    cur = 0
                else:
                    cur = max(min(arr[i], arr[j]), max(cur, arr[j-1]))
                    dd[(i, j)] = cur

        t = [(idx, e) for idx, e in enumerate(arr)]
        t.sort(key=lambda x: x[1])

        res = {}
        for i in range(len(t)):
            res[t[i][0]] = 1
            for j in range(0, i):
                t1, t2 = min(t[j][0], t[i][0]), max(t[j][0], t[i][0])
                if t2 - t1 <= d and t[i][1] > dd[(t1, t2)]:
                    if 1+res[t[j][0]] > res[t[i][0]]:
                        res[t[i][0]] = 1+res[t[j][0]]
        return max([v for k, v in res.items()])


if __name__ == '__main__':
    s = Solution()
    arr = [6, 4, 14, 6, 8, 13, 9, 7, 10, 6, 12]
    d = 2
    print(s.maxJumps(arr, d))

File: leetcode-cpp/weeklycontest/216/problem4.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <stack>
#include <unordered_map>
#include <vector>

using namespace std;

bool compare(const vector<int> &e1, const vector<int> &e2) {
  return max(e1[0] + e2[1], e1[1]) < max(e1[1] + e2[0], e2[1]);
}
class Solution {
public:
  int minimumEffort(vector<vector<int>> &tasks) {
    sort(tasks.begin(), tasks.end(), compare);
    int res = 0;
    int cur = 0;
    // for (auto task : tasks) {
    //   cout << task[0] << " " << task[1] << endl;
    // }
    for (auto task : tasks) {
      if (cur < task[1]) {
        res += (task[1] - cur);
        cur = task[1];
      }
      cur = cur - task[0];
    }
    return res;
  }
};

int main() {
  Solution s;
  // vector<vector<int>> tasks = {{1, 3}, {2, 4}, {10, 11}, {10, 12}, {8, 9}};
  vector<vector<int>> tasks = {{1, 7},  {2, 8},  {3, 9},
                               {4, 10}, {5, 11}, {6, 12}};
  cout << s.minimumEffort(tasks) << endl;
  return 0;
}